lib/timers.js

```diff
@@ -1,52 +1,110 @@
 'use strict';
 
-const Timer = process.binding('timer_wrap').Timer;
+const TimerWrap = process.binding('timer_wrap').Timer;
 const L = require('internal/linkedlist');
-const assert = require('assert').ok;
+const assert = require('assert');
 const util = require('util');
 const debug = util.debuglog('timer');
-const kOnTimeout = Timer.kOnTimeout | 0;
+const kOnTimeout = TimerWrap.kOnTimeout | 0;
 
 // Timeout values > TIMEOUT_MAX are set to 1.
 const TIMEOUT_MAX = 2147483647; // 2^31-1
 
-// IDLE TIMEOUTS
+
+// HOW and WHY the timers implementation works the way it does.
+//
+// Since many parts of Node.js and user applications rely on timeouts, there may
+// be a very large amount of timeouts scheduled at any given time.
+// Therefore, it is very important that the timers implementation is very fast
+// and efficient.
+//
+// Note: It is suggested you first read though the lib/internal/linkedlist.js
+// linked list implementation. How it works is somewhat counter-intuitive to
```

Comment:
"most JavaScript code" is a bit vague. The linked list implementation in `lib/internal/linkedlist.js` is actually quite similar to a lot of textbook linked lists implementations, with a few differences that can make it confusing.

I would suggest removing the "to most JavaScript code" part, and maybe focus on making the `lib/internal/linkedlist.js` implementation easier to understand next (not necessarily as part of this PR). This shouldn't be too risky since it seems `lib/timers.js` is the only user of that (now internal) module.